File: WalletKitJava/WalletKitBRD/src/main/java/com/blockset/walletkit/brd/Key.java
/*
 * Created by Michael Carrara <michael.carrara@breadwallet.com> on 7/1/19.
 * Copyright (c) 2019 Breadwinner AG.  All right reserved.
*
 * See the LICENSE file at the project root for license information.
 * See the CONTRIBUTORS file at the project root for a list of contributors.
 */
package com.blockset.walletkit.brd;

import androidx.annotation.Nullable;

import com.blockset.walletkit.nativex.cleaner.ReferenceCleaner;
import com.blockset.walletkit.nativex.WKKey;
import com.google.common.base.Optional;

import java.util.List;

/* package */
final class Key implements com.blockset.walletkit.Key {

    @Nullable
    static private List<String> wordList;

    /* package */
    static void setDefaultWordList(List<String> wordList) {
        Key.wordList = wordList;
    }

    /* package */
    static List<String> getDefaultWordList() {
        return Key.wordList;
    }

    /* package */
    static boolean isProtectedPrivateKeyString(byte[] keyStringUtf8) {
        return WKKey.isProtectedPrivateKeyString(keyStringUtf8);
    }

    /* package */
    static Optional<Key> createFromPhrase(byte[] phraseUtf8, @Nullable List<String> words) {
        if (words == null) {
            words = Key.wordList;
        }

        if (words == null) {
            return Optional.absent();
        }

        Optional<WKKey> core = WKKey.createFromPhrase(phraseUtf8, words);
        return core.transform(Key::create);
    }

    /* package */
    static Optional<Key> createFromPrivateKeyString(byte[] keyStringUtf8) {
        Optional<WKKey> core = WKKey.createFromPrivateKeyString(keyStringUtf8);
        return core.transform(Key::create);
    }

    /* package */
    static Optional<Key> createFromPrivateKeyString(byte[] keyStringUtf8, byte[] phraseUtf8) {
        Optional<WKKey> core = WKKey.createFromPrivateKeyString(keyStringUtf8, phraseUtf8);
        return core.transform(Key::create);
    }

    /* package */
    static Optional<Key> createFromPublicKeyString(byte[] keyStringUtf8) {
        Optional<WKKey> core = WKKey.createFromPublicKeyString(keyStringUtf8);
        return core.transform(Key::create);
    }

    /* package */
    static Optional<Key> createForPigeon(com.blockset.walletkit.Key key, byte[] nonce) {
        Optional<WKKey> core = WKKey.createForPigeon(from(key).core, nonce);
        return core.transform(Key::create);
    }

    /* package */
    static Optional<Key> createForBIP32ApiAuth(byte[] phraseUtf8, @Nullable List<String> words) {
        if (words == null) {
            words = Key.wordList;
        }

        if (words == null) {
            return Optional.absent();
        }

        Optional<WKKey> core = WKKey.createForBIP32ApiAuth(phraseUtf8, words);
        return core.transform(Key::create);
    }

    /* package */
    static Optional<Key> createForBIP32BitID(byte[] phraseUtf8, int index, String uri, @Nullable List<String> words) {
        if (words == null) {
            words = Key.wordList;
        }

        if (words == null) {
            return Optional.absent();
        }

        Optional<WKKey> core = WKKey.createForBIP32BitID(phraseUtf8, index, uri, words);
        return core.transform(Key::create);
    }

    /* package */
    static Optional<Key> createFromSecret(byte[] secret) {
        Optional<WKKey> core = WKKey.cryptoKeyCreateFromSecret(secret);
        return core.transform(Key::create);
    }

    /* package */
    static Key create(WKKey core) {
        Key key = new Key(core);
        ReferenceCleaner.register(key, core::give);
        return key;
    }

    /* package */
    static Key from(com.blockset.walletkit.Key key) {
        if (key == null) {
            return null;
        }

        if (key instanceof Key) {
            return (Key) key;
        }

        throw new IllegalArgumentException("Unsupported key instance");
    }

    private final WKKey core;

    private Key(WKKey core) {
        this.core = core;
        this.core.providePublicKey(0, 0);
    }

    @Override
    public byte[] encodeAsPrivate() {
        return core.encodeAsPrivate();
    }

    @Override
    public byte[] encodeAsPublic() {
        return core.encodeAsPublic();
    }

    @Override
    public boolean hasSecret() {
        return core.hasSecret();
    }

    @Override
    public byte[] getSecret() {
        return core.getSecret();
    }

    @Override
    public boolean privateKeyMatch(com.blockset.walletkit.Key other) {
        return core.privateKeyMatch(from(other).core);
    }

    @Override
    public boolean publicKeyMatch(com.blockset.walletkit.Key other) {
        return core.publicKeyMatch(from(other).core);
    }

    /* package */
    WKKey getBRCryptoKey() {
        return core;
    }
}

File: WalletKitKotlin/src/jvmMain/kotlin/common/Key.kt
package com.breadwallet.core.common

import com.breadwallet.core.Secret
import com.breadwallet.corenative.cleaner.ReferenceCleaner
import com.breadwallet.corenative.crypto.BRCryptoKey
import kotlinx.io.core.Closeable

actual class Key internal constructor(
        internal val core: BRCryptoKey
) : Closeable {

    init {
        ReferenceCleaner.register(core, ::close)
    }

    internal actual constructor(
            secret: Secret
    ) : this(
            checkNotNull(
                    BRCryptoKey.cryptoKeyCreateFromSecret(secret.u8).orNull()
            )
    )

    actual val hasSecret: Boolean
        get() = core.hasSecret()

    actual val encodeAsPrivate: ByteArray
        get() = checkNotNull(core.encodeAsPrivate())

    actual val encodeAsPublic: ByteArray
        get() = checkNotNull(core.encodeAsPublic())

    actual val secret: Secret
        get() = Secret(core.secret)

    actual fun publicKeyMatch(that: Key): Boolean =
            core.publicKeyMatch(that.core)

    internal actual fun privateKeyMatch(that: Key): Boolean =
            core.privateKeyMatch(that.core)

    override fun close() {
        core.give()
    }

    actual companion object {
        actual var wordList: List<String>? = null
            @Synchronized get
            @Synchronized set

        actual fun isProtectedPrivateKey(privateKey: String): Boolean =
                BRCryptoKey.isProtectedPrivateKeyString(privateKey.toByteArray())

        actual fun createFromPhrase(phrase: String, words: List<String>?): Key? =
                if (words == null && wordList == null) null
                else BRCryptoKey.createFromPhrase(phrase.toByteArray(), words)
                        .orNull()
                        ?.run(::Key)

        actual fun createFromProtectedPrivateKey(privateKey: String, passphrase: String): Key? =
                BRCryptoKey.createFromPrivateKeyString(privateKey.toByteArray(), passphrase.toByteArray())
                        .orNull()
                        ?.run(::Key)

        actual fun createFromPrivateKey(privateKey: String): Key? =
                BRCryptoKey.createFromPrivateKeyString(privateKey.toByteArray())
                        .orNull()
                        ?.run(::Key)

        actual fun createFromPublicKey(string: String): Key? =
                BRCryptoKey.createFromPublicKeyString(string.toByteArray())
                        .orNull()
                        ?.run(::Key)

        actual fun createForPigeonFromKey(key: Key, nonce: ByteArray): Key? =
                BRCryptoKey.createForPigeon(key.core, nonce).orNull()?.run(::Key)

        actual fun createForBIP32ApiAuth(phrase: String, words: List<String>?): Key? =
                if (words == null && wordList == null) null
                else BRCryptoKey.createForBIP32ApiAuth(phrase.toByteArray(), words)
                        .orNull()
                        ?.run(::Key)

        actual fun createForBIP32BitID(phrase: String, index: Int, uri: String, words: List<String>?): Key? =
                if (words == null && wordList == null) null
                else BRCryptoKey.createForBIP32BitID(phrase.toByteArray(), index, uri, words)
                        .orNull()
                        ?.run(::Key)
    }
}
